nav2_collision_monitor/src/collision_monitor_node.cpp

```diff
@@ -0,0 +1,530 @@
+}
+
+void CollisionMonitor::cmdVelInCallback(geometry_msgs::msg::Twist::ConstSharedPtr msg)
+{
+  cmd_vel_in_ = {msg->linear.x, msg->linear.y, msg->angular.z};
+  processMain();
+}
+
+void CollisionMonitor::publishVelocity(const Velocity & vel)
+{
+  std::unique_ptr<geometry_msgs::msg::Twist> cmd_vel_out_msg =
+    std::make_unique<geometry_msgs::msg::Twist>();
+  cmd_vel_out_msg->linear.x = vel.x;
+  cmd_vel_out_msg->linear.y = vel.y;
+  cmd_vel_out_msg->angular.z = vel.tw;
+  // linear.z, angular.x and angular.y will remain 0.0
+
+  cmd_vel_out_pub_->publish(std::move(cmd_vel_out_msg));
+}
+
+bool CollisionMonitor::getParameters(
+  std::string & odom_topic,
+  std::string & cmd_vel_in_topic,
+  std::string & cmd_vel_out_topic)
+{
+  try {
+    auto node = shared_from_this();
+
+    // Initializing polygon topic with empty string.
+    // If it won't be re-writed, polygons won't be published.
+    nav2_util::declare_parameter_if_not_declared(
+      node, "odom_topic", rclcpp::ParameterValue("odom"));
+    odom_topic = get_parameter("odom_topic").as_string();
+    nav2_util::declare_parameter_if_not_declared(
+      node, "cmd_vel_in_topic", rclcpp::ParameterValue("cmd_vel_raw"));
+    cmd_vel_in_topic = get_parameter("cmd_vel_in_topic").as_string();
+    nav2_util::declare_parameter_if_not_declared(
+      node, "cmd_vel_out_topic", rclcpp::ParameterValue("cmd_vel"));
+    cmd_vel_out_topic = get_parameter("cmd_vel_out_topic").as_string();
+
+    nav2_util::declare_parameter_if_not_declared(
+      node, "base_frame_id", rclcpp::ParameterValue("base_footprint"));
+    std::string base_frame_id = get_parameter("base_frame_id").as_string();
+    nav2_util::declare_parameter_if_not_declared(
+      node, "transform_tolerance", rclcpp::ParameterValue(0.1));
+    tf2::Duration transform_tolerance =
+      tf2::durationFromSec(get_parameter("transform_tolerance").as_double());
+    nav2_util::declare_parameter_if_not_declared(
+      node, "data_timeout", rclcpp::ParameterValue(2.0));
+    data_timeout_ =
+      tf2::durationFromSec(get_parameter("data_timeout").as_double());
+    nav2_util::declare_parameter_if_not_declared(
+      node, "simulation_time_step", rclcpp::ParameterValue(0.02));
+    double simulation_time_step = get_parameter("simulation_time_step").as_double();
+
+    nav2_util::declare_parameter_if_not_declared(
+      node, "release_acceleration", rclcpp::ParameterValue(0.1));
+    release_acceleration_ = get_parameter("release_acceleration").as_double();
+
+    // Leave these parameters to be not initialized:
+    // the will cause an error if it will not set
+    nav2_util::declare_parameter_if_not_declared(
+      node, "polygons", rclcpp::PARAMETER_STRING_ARRAY);
+    nav2_util::declare_parameter_if_not_declared(
+      node, "observation_sources", rclcpp::PARAMETER_STRING_ARRAY);
+
+    // Working with polygons parameters
+    std::vector<std::string> polygon_names = get_parameter("polygons").as_string_array();
```

Comment:
Push the polygon and the source creation loops into separate functions that this calls so that this function is shorter

Reply:
Also will help with test coverage

Reply:
Made separate `configurePolygons()`, `configureFootprint()` and `configureSource()` functions in latest update.